src/lib/Simplify.hs

```diff
@@ -551,3 +565,77 @@ hasExceptions expr = do
   case t of
     Nothing -> return $ ExceptionEffect `S.member` effs
     Just _  -> error "Shouldn't have tail left"
+
+-- === Ix simplification ===
+
+data SimpleIxInstance (n::S) =
+  SimpleIxInstance
+    { simpleIxSize            :: (Abs (Nest Decl) LamExpr n)
+    , simpleToOrdinal         :: (Abs (Nest Decl) LamExpr n)
+    , simpleUnsafeFromOrdinal :: (Abs (Nest Decl) LamExpr n)
+    }
+
+instance GenericE SimpleIxInstance where
+  type RepE SimpleIxInstance = (PairE (Abs (Nest Decl) LamExpr)
+                                 (PairE (Abs (Nest Decl) LamExpr)
+                                        (Abs (Nest Decl) LamExpr)))
+  fromE (SimpleIxInstance a b c) = PairE a (PairE b c)
+  toE (PairE a (PairE b c)) = SimpleIxInstance a b c
+
+instance SinkableE SimpleIxInstance
+instance HoistableE SimpleIxInstance
+
+type IxCache = HashMapE (EKey Type) SimpleIxInstance
+
+type MonadIxCache1 (m::MonadKind1) = forall n. MonadState (IxCache n) (m n)
```

Comment:
I'm still not convinced that we should use the `MonadState` interface (i.e. `put`) to update the cache, rather than something like our old `Cat` monad which gets updated based on monoidal extension. Using `State` means that when you try to hoist the cache after you leave a scope you need to query, and attempt to hoist, the entire cache, rather than just the local extension.

Reply:
It actually does not mean that at all! `ScopedT1` implements `MonadState`, but it never even attempts to hoist anything!